core/trace.c

```diff
@@ -47,308 +47,70 @@ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-                    trace->_lf_trace_file
-            );
-            if (items_written != trace->_lf_trace_buffer_size[worker]) {
-                fprintf(stderr, "WARNING: Access to trace file failed.\n");
-                fclose(trace->_lf_trace_file);
-                trace->_lf_trace_file = NULL;
-            }
-        }
-        trace->_lf_trace_buffer_size[worker] = 0;
-    }
-}
-
-/**
- * @brief Flush the specified buffer to a file.
- * @param worker Index specifying the trace to flush.
- */
-void flush_trace(trace_t* trace, int worker) {
-    // To avoid having more than one worker writing to the file at the same time,
-    // enter a critical section.
-    LF_CRITICAL_SECTION_ENTER(GLOBAL_ENVIRONMENT);
-    flush_trace_locked(trace, worker);
-    LF_CRITICAL_SECTION_EXIT(GLOBAL_ENVIRONMENT);
-}
-
-void start_trace(trace_t* trace) {
-    // FIXME: location of trace file should be customizable.
-    trace->_lf_trace_file = fopen(trace->filename, "w");
-    if (trace->_lf_trace_file == NULL) {
-        fprintf(stderr, "WARNING: Failed to open log file with error code %d."
-                "No log will be written.\n", errno);
-    }
-    // Do not write the trace header information to the file yet
-    // so that startup reactions can register user-defined trace objects.
-    // write_trace_header();
-    trace->_lf_trace_header_written = false;
-
-    // Allocate an array of arrays of trace records, one per worker thread plus one
-    // for the 0 thread (the main thread, or in an single-threaded program, the only
-    // thread).
-    trace->_lf_number_of_trace_buffers = _lf_number_of_workers + 1;
-    trace->_lf_trace_buffer = (trace_record_t**)malloc(sizeof(trace_record_t*) * trace->_lf_number_of_trace_buffers);
-    for (int i = 0; i < trace->_lf_number_of_trace_buffers; i++) {
-        trace->_lf_trace_buffer[i] = (trace_record_t*)malloc(sizeof(trace_record_t) * TRACE_BUFFER_CAPACITY);
-    }
-    // Array of counters that track the size of each trace record (per thread).
-    trace->_lf_trace_buffer_size = (int*)calloc(sizeof(int), trace->_lf_number_of_trace_buffers);
-
-    trace->_lf_trace_stop = 0;
-    LF_PRINT_DEBUG("Started tracing.");
-}
-
-void tracepoint(
-        trace_t* trace,
-        trace_event_t event_type,
+void call_tracepoint(
+        int event_type,
         void* reactor,
-        tag_t* tag,
+        tag_t tag,
         int worker,
         int src_id,
         int dst_id,
         instant_t* physical_time,
         trigger_t* trigger,
         interval_t extra_delay,
-        bool is_interval_start
+        bool is_interval_start  // FIXME: Unused argument. It originally was a micro-optimization which I suspect is not needed, but if there is (statistically significant) empirical evidence that it is needed, then we should use it. Note that without a memory barrier, even its original use may have had no effect (i.e., it could have generated the same assembly code anyway).
```

Comment:
Edward has already explained why we do this. I couldn't find the comment where he explained it, but the idea is that when we are flushing the trace buffer, this can create a large delay that will affect our measurement of the length of a time interval. After further thought, I think I have to acknowledge that we don't need to observe this empirically to know that it will happen. But how important is it to us?

The problem is that if we want to address this, we will have to add an API function that prepares to register a tracepoint. This API function would flush the trace buffer if needed. After calling it we could then get the physical time. Alternatively, we could allow the tracepoint function to get the physical time. But the bottom line is that if we want to deal with this, the API will be more complicated. I do not think it is worth it but if anyone disagrees I will happily do what is needed.

Reply:
@edwardalee OK to go ahead and merge as-is?

Reply:
Ok with me to merge this.